app/models/attestation_template.rb

```diff
@@ -21,8 +21,13 @@ class AttestationTemplate < ApplicationRecord
   has_one_attached :signature
 
   validates :footer, length: { maximum: 190 }
-  validates :logo, content_type: ['image/png', 'image/jpg', 'image/jpeg'], size: { less_than: 1.megabytes }
-  validates :signature, content_type: ['image/png', 'image/jpg', 'image/jpeg'], size: { less_than: 1.megabytes }
+
+  FILE_MAX_SIZE = 1.megabytes
+  file_size_validation = Proc.new do
+    { less_than: FILE_MAX_SIZE, message: I18n.t('errors.messages.file_size_out_of_range', file_size_limit: ActiveSupport::NumberHelper.number_to_human_size(FILE_MAX_SIZE)) }
+  end
+  validates :logo, content_type: ['image/png', 'image/jpg', 'image/jpeg'], size: file_size_validation.call
+  validates :signature, content_type: ['image/png', 'image/jpg', 'image/jpeg'], size: file_size_validation.call
```

Comment:
C'est malin, le coup de la Proc anonyme. Cela dit j'ai l'impression aussi que ça oblige à répéter ce code dans tous les modèles concernés.

Je me suis rafraîchi un peu la mémoire sur comment on valide les attachements (j'avais tout oublié), et donc apparemment on utilise la gem `active_storage_validation`. Et il semble que le message d'erreur concernant la taille soit personnalisable directement, sans avoir besoin d'écrire du code custom.

En tout cas, dans la doc de `active_storage_validation`, je lis que [on doit avoir accès à une variable %{min}](https://github.com/igorkasyanchuk/active_storage_validations#internationalization-i18n) pour personnaliser le message.

Du coup, si ça fonctionne (il faudra p-ê mettre à jour la gem ?), on devrait pouvoir écrire :

```
validates :signature, content_type: ['image/png', 'image/jpg', 'image/jpeg'], size: { less_than: 1.megabytes }
```

comme avant, mais juste dans la locale modifier :

```diff
# config/locales/active_storage_validations.fr
fr:
  errors:
    messages:
      content_type_invalid: n’est pas d’un type accepté
-       file_size_out_of_range: "est trop lourde, elle doit faire au plus 200 Mo."
+       file_size_out_of_range: "est trop lourd(e), elle doit faire au plus %{min}."
```

Et d'après la doc ça devrait fonctionner tout seul ? Si ça marche, est-ce que ça résoudrait ton problème ?

Reply:
Effectivement, si la variable %{min} ou quelque chose du genre indiquant le seuil fixé dans le modèle est accessible dans la traduction "file_size_out_of_range" ça simplifierai les choses.

Mais, après avoir mis à jour la [gem](https://github.com/igorkasyanchuk/active_storage_validations) et testé votre proposition, j'ai une erreur m'indiquant que la variable *min* n'est pas accessible à ce niveau.

`I18n::MissingInterpolationArgument in NewAdministrateur::ProceduresController#update
missing interpolation argument :min in "est trop lourd(e), elle doit faire au plus %{min}." (...
`

De ce que j'ai compris de la doc, en fait, les variables sont accessibles sous certaines conditions.. et %{min} serait donc disponible en cas d'erreur sur la validation des dimensions par exemple..

`validates :photos, attached: true, content_type: ['image/png', 'image/jpg', 'image/jpeg'],
                                     dimension: { width: { min: 800, max: 2400 },
                                                  height: { min: 600, max: 1800 }, message: 'is not given between dimension' }
  `
  Dans notre cas, sauf erreur de ma part c'est seulement la variable %{file_size}  qui est accéssible.

Reply:
> C'est malin, le coup de la Proc anonyme. Cela dit j'ai l'impression aussi que ça oblige à répéter ce code dans tous les modèles concernés.
> 

Dans l'esprit du DRY, effectivement on peut refactoriser la proposition initiale pour moins se répéter. 
L'utilisation d'un concerne pourrait aider je pense. qu'en pensez-vous ?

Reply:
Ah oui, effectivement, c'est que pour la taille des images, my bad.

Du coup j'ai deux pistes pour DRY un peu le truc :
1. Soit un concern (ou autre), si c'est ce que tu préfères,
2. Soit on rajoute directement dans la gem `active_record_validations` la possibilité d'accéder à `%{min}` et `%{max}` dans la locale

J'imagine que pour la gem ça serait relativement léger : apparemment le [code du validateur de taille avec les variables disponibles] est assez simple. On pourrait forker la gem, dire au gemfile d'utiliser le fork, et ajouter `%{min}` et `%{max}`. (Et puis ensuite faire une PR au projet principal, mais ça je peux m'en occuper au besoin).

Tu en penses quoi ?

Reply:
Finalement, j'ai opté pour la solution des concerns même si l’exercice de faire un PR sur active_storage_validation semble intéressant.

=> https://github.com/betagouv/demarches-simplifiees.fr/pull/6468/files